types/react-query/index.d.ts

```diff
@@ -1,142 +1,420 @@
-// Type definitions for react-query 0.3
+// Type definitions for react-query 1.1
 // Project: https://github.com/tannerlinsley/react-query
 // Definitions by: Lukasz Fiszer <https://github.com/lukaszfiszer>
 //                 Jace Hensley <https://github.com/jacehensley>
 //                 Matteo Frana <https://github.com/matteofrana>
+//                 Igor Oleinikov <https://github.com/igorbek>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// Minimum TypeScript Version: 3.7
 
-import { ComponentType } from 'react';
+import * as React from 'react';
+import * as _ from 'ts-toolbelt';
 
-// overloaded useQuery function with pagination
-export function useQuery<TResult, TVariables extends object>(
-    queryKey: QueryKey<TVariables>,
-    queryFn: QueryFunction<TResult, TVariables>,
-    options: QueryOptionsPaginated<TResult>
-): QueryResultPaginated<TResult, TVariables>;
+// overloaded useQuery function
+export function useQuery<TResult, TKey extends AnyQueryKey>(
+    queryKey: TKey | false | null | undefined | (() => TKey | false | null | undefined),
+    queryFn: QueryFunction<TResult, TKey>,
+    config?: QueryOptions<TResult>,
+): QueryResult<TResult>;
 
-export function useQuery<TResult, TVariables extends object>(
-    queryKey: QueryKey<TVariables>,
-    queryFn: QueryFunction<TResult, TVariables>,
-    options?: QueryOptions<TResult>
-): QueryResult<TResult, TVariables>;
+export function useQuery<TResult, TKey extends string>(
+    queryKey: TKey | false | null | undefined | (() => TKey | false | null | undefined),
+    queryFn: QueryFunction<TResult, [TKey]>,
+    config?: QueryOptions<TResult>,
+): QueryResult<TResult>;
 
-export type QueryKey<TVariables> = string | [string, TVariables] | false | null | QueryKeyFunction<TVariables>;
-export type QueryKeyFunction<TVariables> = () => string | [string, TVariables] | false | null;
+export function useQuery<TResult, TKey extends AnyQueryKey, TVariables extends AnyVariables>(
+    queryKey: TKey | false | null | undefined | (() => TKey | false | null | undefined),
+    variables: TVariables,
+    queryFn: QueryFunctionWithVariables<TResult, TKey, TVariables>,
+    config?: QueryOptions<TResult>,
+): QueryResult<TResult>;
+
+export function useQuery<TResult, TKey extends string, TVariables extends AnyVariables>(
+    queryKey: TKey | false | null | undefined | (() => TKey | false | null | undefined),
+    variables: TVariables,
+    queryFn: QueryFunctionWithVariables<TResult, [TKey], TVariables>,
+    config?: QueryOptions<TResult>,
+): QueryResult<TResult>;
+
+export function useQuery<TResult, TKey extends AnyQueryKey, TVariables extends AnyVariables = []>({
```

Comment:
Wouldn't it be enough to simply expand this and line 26 to be `TResult | undefined` (maybe `null` too?) ? At least 33 would not necessarily result in undefined, since it's not a falsy value. Then maybe this requires a bit more from the user but at least it shouldn't leave the user in a scenario where things break.

Reply:
Oh nevermind, this is not enough. :disappointed: